File: backend/src/main/java/com/prestashop/web/services/ProductService.java
package com.prestashop.web.services;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import com.prestashop.web.domain.Category;
import com.prestashop.web.domain.Product;
import com.prestashop.web.repository.CartItemRepository;
import com.prestashop.web.repository.CategoryRepository;
import com.prestashop.web.repository.ProductRepository;

import jakarta.persistence.EntityNotFoundException;
import jakarta.transaction.Transactional;

@Service
public class ProductService {

	@Autowired
    private ProductRepository productRepository;
	
	@Autowired
    private CategoryRepository categoryRepository;
	
	@Autowired
    private CartItemRepository cartItemRepository;

    public Optional<Product> findById(Long id) {
        return productRepository.findById(id);
    }

    public List<Product> findAll() {
    	
        return productRepository.findByDeleteAtIsNull();
    }

    public Product save(Product product) {
        return productRepository.save(product);
    }
    
    @Transactional
    public void deleteById(Long id) {
    	
    	Product product = findById(id).orElseThrow( ()->
    			 new IllegalArgumentException("Prodotto non trovato")
    			);
    	
    	product.delete();    	
        productRepository.save(product);
    }

    public Page<Product> findByCategory(Long categoryId, Pageable pageable) {
    	Optional<Category> optionalCategory = categoryRepository.findById(categoryId);
    	Category category = optionalCategory.get();
        return productRepository.findByCategory(category, pageable);
    }
    
    
    public Page<Product> findAllProductShop(Pageable pageable) {
    	
        return productRepository.findByDeleteAtIsNull(pageable);
    }
    
    public Page<Product> findAllProductShopFilter(
            String category, Double minPrice, Double maxPrice, String search, String sortBy, String sortDirection, Pageable pageable) {

            // Specifica per filtrare i prodotti non eliminati
            Specification<Product> spec = Specification.where((root, query, cb) -> cb.isNull(root.get("deleteAt")));

            // Aggiungi filtri
            if (category != null) {
                spec = spec.and((root, query, cb) -> cb.equal(root.get("category").get("title"), category));
            }
            if (minPrice != null) {
                spec = spec.and((root, query, cb) -> cb.greaterThanOrEqualTo(root.get("price"), minPrice));
            }
            if (maxPrice != null) {
                spec = spec.and((root, query, cb) -> cb.lessThanOrEqualTo(root.get("price"), maxPrice));
            }
            if (search != null) {
                spec = spec.and((root, query, cb) -> cb.like(cb.lower(root.get("title")), "%" + search.toLowerCase() + "%"));
            }

            // Ordinamento
            if (sortBy != null && sortDirection != null) {
                pageable = PageRequest.of(
                    pageable.getPageNumber(),
                    pageable.getPageSize(),
                    Sort.by(Sort.Direction.fromString(sortDirection), sortBy)
                );
            }

            
            return productRepository.findAll(spec, pageable);
        }
    
    @Transactional
    public Product createProduct(String title, String description, String imgResources, String imgResources2, String imgResources3, double price, int stock, Long categoryId) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Il titolo del prodotto non può essere vuoto");
        }
        if (price <= 0) {
            throw new IllegalArgumentException("Il prezzo deve essere maggiore di zero");
        }

        Product product = new Product();
        product.setTitle(title);
        product.setDescription(description);
        product.setImgResources(imgResources);
        product.setImgResources2(imgResources2);
        product.setImgResources3(imgResources3);
        product.setPrice(price);
        product.setStock(stock);

        if (categoryId != null) {
            Optional<Category> optionalCategory = categoryRepository.findById(categoryId);
            if (optionalCategory.isEmpty()) {
                throw new EntityNotFoundException("Categoria con ID " + categoryId + " non trovata");
            }
            product.setCategory(optionalCategory.get());
        }

        return productRepository.save(product);
    }
    
    /**
     * Aggiorna un prodotto esistente.
     */
    @Transactional
    public Product updateProduct(Long productId, String title, String description, String imgResources, String imgResources2, String imgResources3, double price, int stock, Long categoryId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new EntityNotFoundException("Prodotto non trovato con ID: " + productId));

        product.setTitle(title);
        product.setDescription(description);
        product.setImgResources(imgResources);
        product.setImgResources2(imgResources2);
        product.setImgResources3(imgResources3);
        product.setPrice(price);
        product.setStock(stock);

        if (categoryId != null) {
            Optional<Category> optionalCategory = categoryRepository.findById(categoryId);
            if (optionalCategory.isEmpty()) {
                throw new EntityNotFoundException("Categoria con ID " + categoryId + " non trovata");
            }
            product.setCategory(optionalCategory.get());
        }

        return productRepository.save(product);
    }
    
    /**
     * Rimuove un prodotto dal database.
     */
    @Transactional
    public void deleteProduct(Long productId) {
        
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new EntityNotFoundException("Prodotto non trovato con ID: " + productId));        
        cartItemRepository.deleteAllByProduct(product);        
        productRepository.delete(product);
    }
    
    public Page<Product> searchProductsByName(String title, Pageable pageable) {
        return productRepository.findByDeleteAtIsNullAndTitleContainingIgnoreCase(title, pageable);
    }
    
    // Metodo per contare il numero di prodotti con deleteAt null.
    public long countProductsNotDeleted() {
        return productRepository.countByDeleteAtIsNull();
    } 
    
}   

File: frontend/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../environment';
import { Observable } from 'rxjs';
import { Product } from '../models/product';
import { RecommendationResponse } from '../models/RecommendationResponse';

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  constructor(private http: HttpClient) {}

  // Ottieni tutti i prodotti
  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${environment.apiUrl}/products`);
  }

  // Ottieni un prodotto per ID
  getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(`${environment.apiUrl}/products/${id}`);
  }

  // Crea un nuovo prodotto
  createProduct(product: { 
    title: string, 
    description: string, 
    imgResources: string, 
    imgResources2: string, 
    imgResources3: string, 
    price: number, 
    stock: number, 
    categoryId: number 
  }): Observable<Product> {
    return this.http.post<Product>(`${environment.apiUrl}/products`, product);
  }

   // Aggiorna un prodotto esistente
   updateProduct(id: number, product: { 
    title: string, 
    description: string, 
    imgResources: string, 
    imgResources2: string, 
    imgResources3: string, 
    price: number, 
    stock: number, 
    categoryId: number 
  }): Observable<Product> {
    return this.http.put<Product>(`${environment.apiUrl}/products/${id}`, product);
  }

  // Elimina un prodotto
  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${environment.apiUrl}/products/${id}`);
  }

  // Ottieni tutti i prodotti per lo Shop
  getAllProductsShop(
    page: number = 0,
    size: number = 12,
    category?: string,
    minPrice?: number,
    maxPrice?: number,
    search?: string,
    sortBy?: string,
    sortDirection?: string
  ): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
  
    if (category) params = params.set('category', category);
    if (minPrice) params = params.set('minPrice', minPrice.toString());
    if (maxPrice) params = params.set('maxPrice', maxPrice.toString());
    if (search) params = params.set('search', search);
    if (sortBy) params = params.set('sortBy', sortBy);
    if (sortDirection) params = params.set('sortDirection', sortDirection);
  
    return this.http.get<any>(`${environment.apiUrl}/products/shop`, { params });
  }

  // Ottieni tutti i prodotti per lo Shop per Admin
  getAllProductsPaginated(page: number, size: number, sort: string, search: string): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    if (search) {
      params = params.set('search', search);
    }

    return this.http.get<any>(`${environment.apiUrl}/products/admin`, { params });
  }

  // Count tutti i prodotti
  getCountProducts(): Observable<number> {
    return this.http.get<number>(`${environment.apiUrl}/products/admin/count`);
  }

  // Shop Assistant
  getRecommendation(query: string): Observable<RecommendationResponse> {    
    const body = { query };
    return this.http.post<RecommendationResponse>(`${environment.apiUrl}/recommendations`, body);
  }
}